### PR TITLE
Add metrics for SecurityPolicy controller

### PR DESCRIPTION
**Pull request by heypnus**

### CODE REVIEW

File: pkg/controllers/securitypolicy_controller.go [hunk 1]
```diff
@@ -44,45 +46,56 @@ type SecurityPolicyReconciler struct {
...
 		if containsString(obj.GetFinalizers(), util.FinalizerName) {
+			metrics.ControllerDeleteTotal.WithLabelValues(RES_TYPE).Inc()
 			if err := r.Service.DeleteSecurityPolicy(obj.UID); err != nil {
+				metrics.ControllerDeleteFailTotal.WithLabelValues(RES_TYPE).Inc()
 				return ctrl.Result{}, err
 			}
 			controllerutil.RemoveFinalizer(obj, util.FinalizerName)
 			if err := r.Client.Update(ctx, obj); err != nil {
+				metrics.ControllerDeleteFailTotal.WithLabelValues(RES_TYPE).Inc()
 				return ctrl.Result{}, err
```
Comment (lxiaopei) [inline pkg/controllers/securitypolicy_controller.go:95]: I think in metrics.go, we can define some functions to update the values, for example, update_counter. In the function, we can also add check for if vmc env.

File: pkg/controllers/securitypolicy_controller.go [hunk 2]
```diff
@@ -189,6 +202,13 @@ func (r *SecurityPolicyReconciler) setupWithManager(mgr ctrl.Manager) error {
...
 func (r *SecurityPolicyReconciler) Start(mgr ctrl.Manager) error {
 	err := r.setupWithManager(mgr)
+	metrics.Register(metrics.ControllerSyncTotal)
+	metrics.Register(metrics.ControllerUpdateTotal)
+	metrics.Register(metrics.ControllerUpdateSuccessTotal)
+	metrics.Register(metrics.ControllerUpdateFailTotal)
+	metrics.Register(metrics.ControllerDeleteTotal)
+	metrics.Register(metrics.ControllerDeleteSuccessTotal)
+	metrics.Register(metrics.ControllerDeleteFailTotal)
 	if err != nil {
```
Comment (lxiaopei) [inline pkg/controllers/securitypolicy_controller.go:212]: I add a func InitializePrometheusMetrics() and only register metrics for vmc env. 
Could we also register all the metrics in that func?

File: pkg/controllers/securitypolicy_controller.go (hunk 1 above)
Comment (heypnus) [reply pkg/controllers/securitypolicy_controller.go:95]: got it. I'll try it.

File: pkg/controllers/securitypolicy_controller.go [hunk 3]
```diff
@@ -41,48 +44,66 @@ type SecurityPolicyReconciler struct {
 	Service *services.SecurityPolicyService
 }
 
+func (r *SecurityPolicyReconciler) counter_inc(counter *prometheus.CounterVec) {
```
Comment (lxiaopei) [inline pkg/controllers/securitypolicy_controller.go:47]: Could we put the counter_inc func into metrics.py? If we add  more controllers in future, we can use them to update metrics value.

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -30,14 +37,68 @@ var (
...
-		prometheus.SummaryOpts{
-			Name:       "rpc_durations_seconds",
-			Help:       "RPC latency distributions.",
-			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+	ControllerSyncTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: MetricNamespace,
+			Subsystem: MetricSubsystem,
+			Name:      ControllerSyncTotalKey,
+			Help:      "Total number K8s create, update and delete events syncronized by NSX operator",
```
Comment (dantingl) [inline pkg/metrics/metrics.go:45]: NSX Operator

File: pkg/metrics/metrics.go [hunk 5]
```diff
@@ -11,10 +11,17 @@ import (
...
 
 const (
-	MetricNamespace = "nsx"
-	MetricSubsystem = "operator"
-	HealthKey       = "health_status"
-	ScrapeTimeout   = 30
+	MetricNamespace                 = "nsx"
+	MetricSubsystem                 = "operator"
+	HealthKey                       = "health_status"
+	ControllerSyncTotalKey          = "controller_sync_total"
```
Comment (dantingl) [inline pkg/metrics/metrics.go:17]: Later if we support more controllers, do we need to define other metrics for sync/update/delete? Or we reuse these ones.

File: pkg/controllers/securitypolicy_controller.go [hunk 6]
```diff
@@ -229,7 +250,11 @@ func (r *SecurityPolicyReconciler) GarbageCollector(cancel chan bool, timeout ti
 				continue
 			}
 			log.V(1).Info("GC collected SecurityPolicy CR", "UID", elem)
-			r.Service.DeleteSecurityPolicy(types.UID(elem))
+			r.counter_inc(metrics.ControllerDeleteTotal)
+			err = r.Service.DeleteSecurityPolicy(types.UID(elem))
+			if err != nil {
+				r.counter_inc(metrics.ControllerDeleteFailTotal)
+			}
```
Comment (lxiaopei) [inline pkg/controllers/securitypolicy_controller.go:257]: Do we need to r.counter_inc(metrics.ControllerDeleteSuccessTotal) if delete successfully?

File: pkg/controllers/securitypolicy_controller.go [hunk 7]
```diff
@@ -32,6 +34,7 @@ var (
 
 const (
 	WCP_SYSTEM_RESOURCE = "vmware-system-shared-t1"
+	RES_TYPE            = "securitypolicy"
```
[the remark this answers is not in the document]
Comment (heypnus) [reply pkg/controllers/securitypolicy_controller.go:37]: yes

File: pkg/controllers/securitypolicy_controller.go (hunk 3 above)
Comment (heypnus) [reply pkg/controllers/securitypolicy_controller.go]: done

File: pkg/controllers/securitypolicy_controller.go (hunk 6 above)
Comment (heypnus) [reply pkg/controllers/securitypolicy_controller.go:253]: fixed, thanks.

File: pkg/metrics/metrics.go (hunk 5 above)
Comment (heypnus) [reply pkg/metrics/metrics.go:17]: we can use these ones with different res_type

File: pkg/metrics/metrics.go (hunk 4 above)
Comment (heypnus) [reply pkg/metrics/metrics.go]: done

File: pkg/controllers/securitypolicy_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dantingl) [reply pkg/controllers/securitypolicy_controller.go:36]: Then how about `METRIC_RES_TYPE`

[the remark this answers is not in the document]
Comment (heypnus) [reply pkg/controllers/securitypolicy_controller.go]: done

**Review by dantingl — APPROVED** [review]

**Review by lxiaopei — APPROVED** [review]

**Pull request merged**